barretenberg/cpp/src/barretenberg/vm/avm/tests/merkle_tree.test.cpp

```diff
@@ -0,0 +1,130 @@
+#include "barretenberg/common/thread.hpp"
+#include "barretenberg/crypto/poseidon2/poseidon2.hpp"
+#include "barretenberg/crypto/poseidon2/poseidon2_params.hpp"
+#include "barretenberg/vm/avm/generated/circuit_builder.hpp"
+#include "barretenberg/vm/avm/generated/flavor.hpp"
+#include "barretenberg/vm/avm/generated/full_row.hpp"
+#include "barretenberg/vm/avm/generated/relations/poseidon2.hpp"
+#include "barretenberg/vm/avm/generated/relations/poseidon2_full.hpp"
+
+#include "barretenberg/vm/avm/trace/gadgets/merkle_tree.hpp"
+#include <cstdint>
+#include <cstdlib>
+#include <gtest/gtest.h>
+
+namespace tests_avm {
+using namespace bb;
+using namespace bb::Avm_vm;
+
+TEST(AvmMerkleTree, shouldCheckMembership)
+{
+
+    using FF = AvmFlavor::FF;
+    constexpr size_t TRACE_SIZE = 1 << 8;
+
+    std::vector<AvmFullRow<FF>> trace(TRACE_SIZE);
+
+    bb::avm_trace::AvmMerkleTreeTraceBuilder merkle_tree_builder;
+    std::cerr << "Generating trace of size " << TRACE_SIZE << "..." << std::endl;
+
+    /* MEMBERSHIP CHECK */
+    // Create a bunch of random elements and hash them.
+    size_t depth = 32;
+    std::vector<FF> sibling_path;
+    FF initial_value = FF::random_element();
+    uint32_t leaf_index = 2000;
+    for (size_t i = 0; i < depth; ++i) {
+        sibling_path.push_back(FF::random_element());
+    }
+    FF root = initial_value;
+    uint32_t leaf_index_copy = leaf_index;
+    for (size_t i = 0; i < depth; ++i) {
+        root = leaf_index_copy % 2 == 0
+                   ? crypto::Poseidon2<crypto::Poseidon2Bn254ScalarFieldParams>::hash({ root, sibling_path[i] })
+                   : crypto::Poseidon2<crypto::Poseidon2Bn254ScalarFieldParams>::hash({ sibling_path[i], root });
+        leaf_index_copy /= 2;
+    }
+
+    // Check membership
+    bool found = merkle_tree_builder.check_membership(0, initial_value, leaf_index, sibling_path, root);
+    ASSERT_TRUE(found);
+    bool not_found = merkle_tree_builder.check_membership(1, initial_value, leaf_index + 1, sibling_path, root);
```

Comment:
I would not name with the "not_" because it is a bit confusing in next line you think that ASSERT_FALSE of not_found actually means "found". The boolean semantic is always found as returned from check_membership but in this case we expect it to be false.